report/3.Theory/Parser.tex

```diff
@@ -0,0 +1,112 @@
+Therefore, another common output type from the parser is the abstract syntax tree.
+
+\textbf{Abstract syntax tree}
+
+The abstract syntax tree is a sparse version of the parse tree. 
+It still contains all the information needed to understand the structure of the program, but in a more compressed way. 
+For instance parentheses can be omitted in the abstract syntax tree.
+
+Recall the parse tree from figure \ref{fig:AmbiguousGrammarEx1} a corresponding abstract syntax tree can bee seen on figure \ref{fig:ast} 
+
+\begin{figure}[H]
+	\centering
+	\includegraphics[width=\textwidth/2+\textwidth/4]{3.Theory/images/ast.png}
+	\caption{
+		Abstract syntax tree for parse tree in figure \ref{fig:AmbiguousGrammarEx1}.
+	}
+	\label{fig:ast}
+\end{figure}
+
+%\subsection{Grammar}
+%
+%The parser is built upon the CFG defined for the language. 
+%It is the responsibility of the CFG to tell the parser how to parse the language. 
+%For instance, when writing a top down parser, first and follow sets can be generated on behalf of the CFG. 
+%These will help determine what should be parsed next. 
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% I guess it is not really necessary to talk about CFG, when we already have a big section about it.
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+\subsection{Contextual analysis}
+
+Apart from syntactical analysis the parser is also concerned with contextual analysis. 
+The contextual analysis ensures that types are correct, immutable variables are not reassigned and other contextual things that cannot be inferred from the CFG due to its context free nature. 
+The major part of the contextual analysis is the type checking.
+
+\textbf{Type checking}
+
+The type checking part of the parser often utilizes a symbol table.
+
+INSERT JONAS SYMBOL TABLE STUFF HERE.
+
+It is also the responsibility of this part of the parser to provide any relevant type information that may be needed in order to proceed with compilation in later phases. 
+There are several ways to inform the later stages of the compiler about the needed type information. 
+One solution is to store and pass the symbol table to the later stages of the compiler. 
+
+Another way is to decorate the abstract syntax tree with the needed type information. 
+This is called a decorated syntax tree, and will in most cases be the final output of the parser.
+An example of decorating an abstract syntax tree could be when using implicit casts. 
+Most languages have this feature between integers and floating point numbers. 
+So if the user were to do arithmetic with an integer and a floating point number, the integer would then be implicitly casted to a floating point number.
+
+\subsection{Single parse and multi parse}
+
+Depending on the language, a single run through the passer may not be enough to validate the input or generate the output. 
+For instance if the language allows calling of functions or usage of classes that has not yet been declared. 
```

Comment:
comma after 'instance'
should maybe be " allows the calling of functions or the usage of classes"